### DISCUSSION TITLE
Kernel dies after the .run

### DISCUSSION THREAD

**Discussion by EvgenyUgolkov** [PMEAL/OpenPNM#2683] [Q&A] [answered]
Dear Team, good day. Many thanks for your previous response, it helped me to advance a little bit further, but i stuck again. Could you help me?

My goal is to calculate the absolute permeability of the 3D Micro-CT image of sandstone, which i upload from the .TIF file (i attached this file in the .zip archive). The image has a size of 400x400x400 voxels, what seems to be not very large. I use the following code, and after .run my Jupyter Notebook kernel dies after 1 second.

Dear Team, could you tell me if i am doing anything wrong here? Also, maybe you have some recommendations how i can make this process better? Maybe i should add or optimize anything for better/faster results?

Here the code and the file in zip archive go

With respect, Evgeny

[berea.zip](https://github.com/PMEAL/OpenPNM/files/10611553/berea.zip)

```ruby
import torch
import openpnm as op
import porespy as ps
import numpy as np
import skimage.io as skio 
import matplotlib.pyplot as plt

# import 3D bimary image (0 and 255 values) into the tensor and normalize it to 0 and 1 values

Image_3D = torch.tensor(skio.imread('berea.tif', plugin="tifffile")).unsqueeze(0).float()/255
voxel_size = 5.345e-6 # 5.345 micro meters

# Transfer 0 and 1 into the PoreSpy convention where 0 = grains and 1 = pores. Check with porosity function 
# if everything is correct. Transfer the torch tensor into numpy array

Image_3D = 1 - Image_3D
print('Image_3D porosity', ps.metrics.porosity(Image_3D))
Image_3D = Image_3D[0].numpy()

# Extract the network with the PoreSpy and assign boundaries

snow_output = ps.networks.snow2(Image_3D, voxel_size=voxel_size)
snow_output = ps.networks.label_boundaries(snow_output.network, labels=[['left', 'right'], ['front', 'back'], ['top', 'bottom']])

# Put the extracted network into the OpenPNM

pn = op.io.network_from_porespy(snow_output)

# Determine the parameters of the network

pn['pore.diameter'] = pn['pore.equivalent_diameter']
pn['throat.diameter'] = pn['throat.inscribed_diameter']
pn.add_model_collection(op.models.collections.geometry.pyramids_and_cuboids)
pn.add_model(propname='throat.length', model=op.models.geometry.throat_length.cubes_and_cuboids)
pn.add_model(propname='throat.lens_volume', model=op.models.geometry.throat_volume.lens)
pn.add_model(propname='throat.total_volume', model=op.models.geometry.throat_volume.cuboid)
pn.add_model(propname='throat.volume', model=op.models.misc.difference, props=['throat.total_volume', 'throat.lens_volume'])

# Check the network for isolated and disconnected pores. Delete (trim) these pores

h = op.utils.check_network_health(pn)
op.topotools.trim(network=pn, pores=h['isolated_pores'])
h = op.utils.check_network_health(pn)
op.topotools.trim(network=pn, pores=h['disconnected_pores'])

# Determine the properties of the water and determine conductance

water = op.phase.Water(network=pn)
pn.add_model(propname='throat.hydraulic_size_factors', model=op.models.geometry.hydraulic_size_factors.pyramids_and_cuboids)
water.add_model(propname='throat.hydraulic_conductance',model=op.models.physics.hydraulic_conductance.hagen_poiseuille)

# Determine the Stocks Flow and boundary conditions

flow = op.algorithms.StokesFlow(network=pn, phase=water)
Pin, Pout = (101_325, 0)
flow.set_value_BC(pores=pn.pores('left'), values=Pin)
flow.set_value_BC(pores=pn.pores('right'), values=Pout)

# Check that before .run everything seems to work fine

print('Till this point no errors')

# Run the simulaiton
flow.run()
```

![image](https://user-images.githubusercontent.com/92923525/216832057-7469b9d1-5851-4659-a878-47dd127a7c0e.png)

**Comment by jgostick** [comment, marked as answer]
Can you clarify how far the script gets before crashing?  Break it into separate cells and see which is the problem?

**Reply by JahanzebTariq** [reply]
![image](https://user-images.githubusercontent.com/99289270/216881354-ebb939cc-77ab-46bf-91cc-c56d210d7998.png)
is this right..?

**Reply by EvgenyUgolkov** [reply]
@jgostick , good day. Everything works fine till the very last line: flow.run
@JahanzebTariq , Hi) i tried both: just 101325 and 101_325. The latter was taken from the example on the website

**Reply by JahanzebTariq** [reply]
@EvgenyUgolkov .. can you give me the link of the example?

**Reply by EvgenyUgolkov** [reply]
@JahanzebTariq just the Basic Usage on the openpnm website. Here is the link and the point from where i got it

https://openpnm.org/basic_usage.html

![image](https://user-images.githubusercontent.com/92923525/216900447-0669b53e-d43f-4f23-b34c-e0c8066008ff.png)

**Reply by EvgenyUgolkov** [reply]
@JahanzebTariq also maybe this example will be usefull for you. This one was implemented with a new version 3.0 and contains new syntax. I took many lines of my code from here. It helps to understand the structure and sequence of actions

https://github.com/PMEAL/OpenPNM/discussions/2601

**Reply by jgostick** [reply]
@JahanzebTariq thanks for taking a look.  The under-score is just a python trick to help with reading of large numbers. 

@EvgenyUgolkov, you script looks ok to me, so not sure yet what to recommend.  Can you try running everything exactly the same, but on a 10x10x10 cubic network?  (pn = op.network.Cubic([10, 10, 10], spacing=1e-5))

**Reply by JahanzebTariq** [reply]
@jgostick 

`water.add_model(propname='throat.hydraulic_conductance',model=op.models.physics.hydraulic_conductance.hagen_poiseuille)`

is this step necessary, I mean is hydraulic conductance not automatically calculated when doing op.model.collections.physics?

**Reply by jgostick** [reply]
You are correct, that model is part of the standard physics collection

**Reply by JahanzebTariq** [reply]
so is there a need to mention the equation as mentioned in the above coding or using the standard library is enough?

**Reply by jgostick** [reply]
It should not have any effect...it will just overwrite the model that is already there.

**Reply by EvgenyUgolkov** [reply]
@jgostick Professor, good day. I found the code with Berea Sandstone which you sent to JahanzebTariq 2 month ago, it helped. Many thanks for support!

**Reply by jgostick** [reply]
Can you remind me what the solution was?

**Reply by EvgenyUgolkov** [reply]
@jgostick it was not the solution for the dying kernel, but it was a useful example which helped me to understand the sequence of actions. Somehow by chance i noticed that if i first run the simple example of absolute permeability from the github, (here is it)

![image](https://user-images.githubusercontent.com/92923525/217461850-ba534033-f671-43ec-a302-6575d62dff04.png)

and after it run my code, then kernel does not die. However, if i don't run this simple example first, kernel dies on my example. I don't know why is it, but just use. Maybe you know?
